src/Microsoft.OData.Core/UriParser/Parsers/FunctionOverloadResolver.cs

```diff
@@ -217,45 +216,58 @@ internal static bool ResolveOperationFromList(string identifier, IEnumerable<str
                     throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_SegmentDoesNotSupportKeyPredicates(identifier));
                 }
 
-                matchingOperation = candidateMatchingOperations.Single();
+                matchingOperation = candidatesMatchingOperations.Count > 0 ? candidatesMatchingOperations[0] : null;
                 return true;
             }
 
-            if (foundActionsWhenLookingForFunctions.Count > 0)
+            if (foundActionsWhenLookingForFunctions != null && foundActionsWhenLookingForFunctions.Count > 0)
             {
                 throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_SegmentDoesNotSupportKeyPredicates(identifier));
             }
 
             // If more than one overload matches, try to select based on optional parameters
-            if (candidateMatchingOperations.Count() > 1)
+            if (candidatesMatchingOperations.Count > 1)
             {
-                candidateMatchingOperations = candidateMatchingOperations.FindBestOverloadBasedOnParameters(parameterNames);
+                candidatesMatchingOperations = candidatesMatchingOperations.FindBestOverloadBasedOnParameterCount(parameterNames.Count);
             }
 
-            if (candidateMatchingOperations.Count() > 1)
+            if (candidatesMatchingOperations.Count > 1)
             {
                 throw new ODataException(ODataErrorStrings.FunctionOverloadResolver_NoSingleMatchFound(identifier, string.Join(",", parameterNames.ToArray())));
             }
 
-            matchingOperation = candidateMatchingOperations.SingleOrDefault();
+            matchingOperation = candidatesMatchingOperations.Count > 0 ? candidatesMatchingOperations[0] : null;
             return matchingOperation != null;
         }
 
         internal static bool HasAny<T>(this IEnumerable<T> enumerable) where T : class
         {
-            IList<T> list = enumerable as IList<T>;
-            if (list != null)
+            if (enumerable != null && enumerable.GetEnumerator().MoveNext())
             {
-                return list.Count > 0;
+                return true;
             }
 
-            T[] array = enumerable as T[];
-            if (array != null)
+            return false;
+        }
+
+        internal static bool CountGreaterThan<T>(this IEnumerable<T> enumerable, int count) where T : class
+        {
+            int currentCount = 0;
+            foreach (T element in enumerable)
             {
-                return array.Length > 0;
+                if (element == null)
+                {

```

Comment:
Should we skip null elements? The previous logic that was using Count() > 1 didn't really skip null elements.

Reply:
I expect none of the elements to be null. If they are null, we should not be counting those anyway. I explicitly had to add this check as the element value was not being used otherwise. 

---
In reply to: [370402281](https://github.com/OData/odata.net/pull/1602#discussion_r370402281) [](ancestors = 370402281)